File: java/src/main/java/org/caseontology/InvestigationBuilder.java
package org.caseontology;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Scenario + evidence → profile-aware CaseGraph with inline critique. */
public final class InvestigationBuilder {
    public static final class CritiqueFinding {
        public final String severity;
        public final String message;
        public final String path;

        public CritiqueFinding(String severity, String message, String path) {
            this.severity = severity;
            this.message = message;
            this.path = path;
        }
    }

    private final String scenario;
    private final String profileId;
    private final CaseGraph graph;
    private final List<CritiqueFinding> findings = new ArrayList<>();

    public InvestigationBuilder(String scenario) {
        this(scenario, null, "http://example.org/kb/");
    }

    public InvestigationBuilder(String scenario, String profileId) {
        this(scenario, profileId, "http://example.org/kb/");
    }

    public InvestigationBuilder(String scenario, String profileId, String kbPrefix) {
        this.scenario = scenario == null ? "" : scenario;
        this.profileId = CompositionProfiles.resolve(profileId, this.scenario);
        Map<String, String> extra = new LinkedHashMap<>();
        if (CompositionProfiles.requiresCac(this.profileId)) {
            extra.put("cac-core", "https://cacontology.projectvic.org/core#");
            extra.put("cacontology", "https://cacontology.projectvic.org#");
        }
        this.graph = extra.isEmpty() ? new CaseGraph(kbPrefix) : new CaseGraph(kbPrefix, extra);
    }

    public String getProfileId() { return profileId; }
    public CaseGraph getGraph() { return graph; }

    public Object addFile(String fileName, List<String[]> hashes) {
        List<String[]> list = hashes == null ? Collections.emptyList() : hashes;
        if (list.isEmpty()) {
            findings.add(new CritiqueFinding(
                "error", fileName + ": " + profileId + " requires ContentDataFacet hashes", fileName));
        }
        return CompositionHelpers.fileWithContentHashes(graph, fileName, list);
    }

    public CompositionHelpers.CsamEvidence addCsamEvidence(String fileName, List<String[]> hashes) {
        List<String[]> list = hashes == null ? Collections.emptyList() : hashes;
        if (list.isEmpty()) {
            findings.add(new CritiqueFinding(
                "error", fileName + ": CSAM evidence must carry hashes", fileName));
        }
        return CompositionHelpers.modelCsamEvidence(graph, fileName, list);
    }

    public CompositionHelpers.ToolRun addToolRun(String toolName, String actionName, String toolVersion) {
        if (toolVersion == null || toolVersion.isEmpty()) {
            findings.add(new CritiqueFinding("warning", "Tool " + toolName + " has no version", toolName));
        }
        return CompositionHelpers.modelToolRun(graph, toolName, actionName, toolVersion);
    }

    public CaseGraph build() { return graph; }

    public List<CritiqueFinding> critique() { return Collections.unmodifiableList(findings); }
}

File: csharp/CaseUco/InvestigationBuilder.cs
// Profile-aware InvestigationBuilder — additive topology surface.

using System;
using System.Collections.Generic;

namespace CaseUco
{
    public sealed class CritiqueFinding
    {
        public string Severity { get; set; }
        public string Message { get; set; }
        public string Path { get; set; }
        public string RuleId { get; set; }
        public string FindingId { get; set; }
        public string RecommendedChange { get; set; }
        public string RepairHint { get; set; }
    }

    /// <summary>
    /// Scenario + evidence → profile-aware <see cref="CaseGraph"/> with inline critique.
    /// </summary>
    public sealed class InvestigationBuilder
    {
        public string Scenario { get; }
        public string ProfileId { get; }
        public CaseGraph Graph { get; }
        private readonly List<CritiqueFinding> _findings = new List<CritiqueFinding>();

        public InvestigationBuilder(string scenario, string profileId = null, string kbPrefix = "http://example.org/kb/")
        {
            Scenario = scenario ?? "";
            ProfileId = CompositionProfiles.Resolve(profileId, Scenario);
            var extra = new Dictionary<string, string>();
            if (CompositionProfiles.RequiresCac(ProfileId))
            {
                extra["cac-core"] = "https://cacontology.projectvic.org/core#";
                extra["cacontology"] = "https://cacontology.projectvic.org#";
            }
            Graph = new CaseGraph(kbPrefix, extra.Count > 0 ? extra : null);
        }

        public object AddFile(string fileName, IEnumerable<KeyValuePair<string, string>> hashes = null)
        {
            var list = hashes == null ? new List<KeyValuePair<string, string>>() : new List<KeyValuePair<string, string>>(hashes);
            if (list.Count == 0)
            {
                _findings.Add(new CritiqueFinding
                {
                    Severity = "error",
                    Message = fileName + ": " + ProfileId + " requires ContentDataFacet hashes",
                    Path = fileName,
                });
            }
            return CompositionHelpers.FileWithContentHashes(Graph, fileName, list);
        }

        public CsamEvidence AddCsamEvidence(string fileName, IEnumerable<KeyValuePair<string, string>> hashes)
        {
            var list = hashes == null ? new List<KeyValuePair<string, string>>() : new List<KeyValuePair<string, string>>(hashes);
            if (list.Count == 0)
            {
                _findings.Add(new CritiqueFinding
                {
                    Severity = "error",
                    Message = fileName + ": CSAM evidence must carry hashes",
                    Path = fileName,
                });
            }
            return CompositionHelpers.ModelCsamEvidence(Graph, fileName, list);
        }

        public ToolRun AddToolRun(string toolName, string actionName, string toolVersion = null)
        {
            if (string.IsNullOrEmpty(toolVersion))
            {
                _findings.Add(new CritiqueFinding
                {
                    Severity = "warning",
                    Message = "Tool " + toolName + " has no version",
                    Path = toolName,
                });
            }
            return CompositionHelpers.ModelToolRun(Graph, toolName, actionName, toolVersion);
        }

        public CaseGraph Build() => Graph;

        public IReadOnlyList<CritiqueFinding> Critique() => _findings;
    }
}
